### DISCUSSION TITLE
How should we expect $collection->onEachSide->links() to work?

### DISCUSSION THREAD

**Discussion by geoff-maddock** [laravel/framework#39160] [General]
I was looking into a solution for limiting the number of pages displayed and came across 

https://laravel.com/docs/8.x/pagination#adjusting-the-pagination-link-window

I tried it out by using ```{!! $events->onEachSide(2)->links() !!}```

However, what I see doesn't seem to produce the results I would expect.

Up through the 6th page of results here, it shows 8 pages, then ... and two at the end:
![pagesoneachside](https://user-images.githubusercontent.com/55493/136713284-38fb6739-2cfa-48b3-86b3-9831be9e8623.png)

Only once I get to the 7th page does it start paginating how I might expect:
![Screenshot (7)](https://user-images.githubusercontent.com/55493/136713322-5649d3fc-defc-4d3c-96ed-e4c78c78b7af.png)

Is there any method to alter this initial paging window (1-7?)

**Comment by meduzen** [comment]
Hi, I ran into the same situation. Here are my findings, and a workaround.

Unfortunately `onEachSide` is not meant to be used with a value below 3, because the logic computing the amount of links to show when you are at the beginning or the end of the slider is hardcoded:

https://github.com/laravel/framework/blob/e2fdcd734bbf4d7bf254faa17ad8ad601b6aa24b/src/Illuminate/Pagination/UrlWindow.php#L73-L136

In order to keep low maintenance of the pagination template, what I’ve done is the following (minimal version based on Laravel [`pagination.blade.php`](https://github.com/laravel/framework/blob/12.x/src/Illuminate/Pagination/resources/views/default.blade.php)):


The parts added are:

- the code inside the `@php` directive
- the `@if` directive using `$has_pages_out_of_bounds`

```diff
@if ($paginator->hasPages())
    <nav>
        <ul class="pagination">
            {{-- Previous Page Link --}}
            @if ($paginator->onFirstPage())
                <li class="disabled" aria-disabled="true" aria-label="@lang('pagination.previous')">
                    <span aria-hidden="true">&lsaquo;</span>
                </li>
            @else
                <li>
                    <a href="{{ $paginator->previousPageUrl() }}" rel="prev" aria-label="@lang('pagination.previous')">&lsaquo;</a>
                </li>
            @endif

+           @php
+           /**
+            * When there is a separator, `$elements` has up to 5 keys
+            * (with `0` and `4` always present, and `1`, `2` and
+            * `3` sometimes missing):
+
+            * - `$elements[0]`: an array with the first pages
+            * - `$elements[1]`: separator (`...`)
+            * - `$elements[2]`: an array with pages surrounded by separators
+            * - `$elements[3]`: separator (`...`)
+            * - `$elements[4]`: an array with the last pages
+            *
+            * Since `onEachSide($number)` isn’t intended to be < 3, we
+            * need to know which pages are out of bound. We can not
+            * mutate `$elements`: it would change its keys. The
+            * exclusion is done in the template `@foreach()`.
+            */
+           $has_pages_out_of_bounds = count($elements) > 1 && !isset($elements[2])
+               && (
+                   count($elements[0]) > $paginator->onEachSide
+                   || count($elements[4]) > $paginator->onEachSide
+               );
+           @endphp

            {{-- Pagination Elements --}}
            @foreach ($elements as $element)
                {{-- "Three Dots" Separator --}}
                @if (is_string($element))
                    <li class="disabled" aria-disabled="true"><span>{{ $element }}</span></li>
                @endif

                {{-- Array Of Links --}}
                @if (is_array($element))
                    @foreach ($element as $page => $url)
                        @if ($page == $paginator->currentPage())
                            <li class="active" aria-current="page"><span>{{ $page }}</span></li>
                        @else
+                           {{-- Skip pages out of wished boundaries. --}}
+                           @if (
+                               $has_pages_out_of_bounds
+                               && (
+                                   ($index == 0 && $page > $paginator->currentPage() + $paginator->onEachSide)
+                                   || ($index == 4 && $page < $paginator->currentPage() - $paginator->onEachSide)
+                               )
+                           )
+                               @continue
+                           @endif
                            <li><a href="{{ $url }}">{{ $page }}</a></li>
                        @endif
                    @endforeach
                @endif
            @endforeach

            {{-- Next Page Link --}}
            @if ($paginator->hasMorePages())
                <li>
                    <a href="{{ $paginator->nextPageUrl() }}" rel="next" aria-label="@lang('pagination.next')">&rsaquo;</a>
                </li>
            @else
                <li class="disabled" aria-disabled="true" aria-label="@lang('pagination.next')">
                    <span aria-hidden="true">&rsaquo;</span>
                </li>
            @endif
        </ul>
    </nav>
@endif
```

Basically the code instructs to skip outputting HTML for the pages out of bounds, and the condition runs only when we are sure there is at least one page out of bound, so it should not have bad performances.

This gives the following before/after for `onEachSide(2)`:

| before | after |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/d6443f82-370f-40e6-b254-d6e16596e88f) | ![image](https://github.com/user-attachments/assets/0807de2b-8fe4-4db4-86f1-1ea2683fd5c1) |
| ![image](https://github.com/user-attachments/assets/bb4e1004-7813-4ffa-ab0c-aecb01a8a215) | ![image](https://github.com/user-attachments/assets/7f87b788-e73d-4a0b-9053-429ecb268879) |
| ![image](https://github.com/user-attachments/assets/a84e7aba-2821-4aff-92bf-ce1595ea9e65) | ![image](https://github.com/user-attachments/assets/a84e7aba-2821-4aff-92bf-ce1595ea9e65) |
| ![image](https://github.com/user-attachments/assets/a2d4d8b1-7a72-4a95-b0c6-08ceb9d7e8d6) | ![image](https://github.com/user-attachments/assets/a2d4d8b1-7a72-4a95-b0c6-08ceb9d7e8d6) |
| ![image](https://github.com/user-attachments/assets/8ad1f6fd-4978-476c-a882-e442ded15101) | ![image](https://github.com/user-attachments/assets/589835c6-5971-4c35-aa8b-001d83f84851) | 

The downside is, when you don’t have separators (11 pages or less using `onEachSide(2)`), this code doesn’t change anything:
![image](https://github.com/user-attachments/assets/da6da5b9-e68b-44bf-9ff8-46f354c4f3b1)

Changing this would require more code. Maybe I’ll do it at some point, but for now it’s acceptable like this in my project.
